### PR TITLE
Align Kimi BF16 column linear under accuracy gate

### PR DESCRIPTION
**Pull request by zrr1999**
### PR Category
Operator Mechanism

### PR Types
Bug fixes

### Description
This PR narrows the Kimi accuracy-compatible BF16 column-linear alignment change to the experimental ColumnParallelLinear path only.

- Add an explicit `use_accuracy_compatible` argument to `column_sequence_parallel_linear`.
- In TP=1 BF16 only, route through `paddle.nn.functional.linear` to match the Megatron-Core no-TE behavior observed for Kimi shared-expert fc1.
- Keep the default `fused_linear` path unchanged unless `use_accuracy_compatible=True`.
- Add a single-card regression test for the accuracy gate, default fused path, dtype guard, and `ColumnParallelLinear.forward` config propagation.

Validation:
- `python -m py_compile src/paddlefleet/tensor_parallel/layers.py tests/single_card_tests/tensor_parallel/test_column_sequence_parallel_accuracy.py`
- `pre-commit run --files tests/single_card_tests/tensor_parallel/test_column_sequence_parallel_accuracy.py src/paddlefleet/tensor_parallel/layers.py --show-diff-on-failure`
- Kimi 2-layer no-TE forward-loss comparisons reached 20-step and 100-step `BITWISE_PASS` in the repro workspace.

### CODE REVIEW

File: src/paddlefleet/transformer/transformer_config.py [hunk 1]
```diff
@@ -984,6 +988,23 @@ def __post_init__(self):
...
+            self.high_precision_rope = False
+            self.bf16 = True
+            self.attention_softmax_in_fp32 = True
+            self.masked_softmax_fusion = False
+            self.apply_rope_fusion = False
+            self.qk_norm_fusion = False
+            self.moe_topk_fusion = False
+            self.routing_map_fusion = False
+            self.fused_linear_ce_loss_chunk = 0
+            os.environ["FLAGS_use_accuracy_compatible_kernel"] = "1"
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py]: P1 **优先级：P1**

这里在构造一个 `use_accuracy_compatible=True` 的 config 时会永久写入进程级环境变量。这个 flag 现在被多个运行时路径动态读取（例如 fused softmax、language loss、MoE dispatcher/hyper_connection），因此同一 Python 进程里只要先创建过 accuracy config，之后创建的普通 config 即使 `use_accuracy_compatible=False` 也会被这些 env-gated 路径污染，默认训练/推理路径不再可恢复。请不要在 `TransformerConfig.__post_init__` 里修改 `os.environ`，而是把 config gate 显式传到需要的模块；如果还要兼容外部手动设置 env，可以把 env 只作为用户显式开关读取。

可以按这个方向收敛：

```python
def accuracy_compatible_enabled(config=None):
    return bool(getattr(config, "use_accuracy_compatible", False)) or (
        os.environ.get("FLAGS_use_accuracy_compatible_kernel", "0") == "1"
    )
```

然后由持有 `config` 的调用点使用该 helper，而不是从 config 构造阶段反向写全局环境变量。

File: src/paddlefleet/tensor_parallel/layers.py [hunk 2]
```diff
@@ -352,15 +352,25 @@ class LinearWithFrozenWeight(paddle.autograd.Function):
...
         ctx.tp_group = tp_group
-        output = paddle.matmul(input, weight)
-
         if bias is not None:
-            output = output + bias
+            output = paddle.nn.functional.linear(input, weight, bias)
+        elif use_accuracy_compatible:
+            output = paddle.nn.functional.linear(input, weight)
+        else:
+            output = paddle.matmul(input, weight)
```
Comment (risemeup1111) [inline src/paddlefleet/tensor_parallel/layers.py]: P1 **优先级：P1**

这里把 frozen-weight 的 bias 分支无条件切到 `F.linear`，即使 `use_accuracy_compatible=False` 也会改变默认路径的 GEMM/融合选择；这和 PR 描述的“default fused_linear path unchanged unless use_accuracy_compatible is enabled”不一致，也会让普通 BF16/FP16 冻结权重层出现数值或性能回归。请只在 accuracy gate 下使用 `F.linear`，默认路径保持原来的 `matmul + bias`。

```suggestion
        if use_accuracy_compatible:
            output = paddle.nn.functional.linear(input, weight, bias)
        else:
            output = paddle.matmul(input, weight)
            if bias is not None:
                output = output + bias
```

File: src/paddlefleet/tensor_parallel/layers.py [hunk 3]
```diff
@@ -352,15 +352,25 @@ class LinearWithFrozenWeight(paddle.autograd.Function):
...
     @staticmethod
-    def forward(ctx, input, weight, bias, allreduce_dgrad, tp_group):
+    def forward(
+        ctx,
+        input,
+        weight,
+        bias,
+        allreduce_dgrad,
+        tp_group,
+        use_accuracy_compatible=False,
```
Comment (risemeup1111) [inline src/paddlefleet/tensor_parallel/layers.py]: P1 **优先级：P1**

这个新参数让 frozen-weight 前向也能走 accuracy `F.linear`，但同一个类的 backward 仍直接使用保存的 `weight.t()` 做 dgrad。PR 在 trainable linear backward 已经处理了同类问题：BF16 前向接 CE/上游 FP32 梯度时，`grad_output` 可能是 float32 而冻结权重仍是 bfloat16，Paddle matmul 需要同 dtype，会在 LoRA/冻结 base weight 路径反向时报 dtype mismatch。请同步修正 frozen-weight backward：

```python
dgrad_weight = (
    weight.cast(grad_output.dtype)
    if weight.dtype != grad_output.dtype
    else weight
)
grad_input = grad_output.matmul(dgrad_weight.t())
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已完成首轮审查。发现需要先修复的 P1 问题，细节已放在行内评论，主要涉及 accuracy gate 的进程级状态泄漏，以及 frozen-weight 线性层默认路径和反向 dtype 处理的回归风险。CI 当前通过。

P1 **优先级：P1**
- 非行级：GitHub mergeability 当前仍显示存在 merge conflicts；请同步 develop 并解决冲突后再请求复查。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/mlp.py [hunk 4]
```diff
@@ -210,7 +214,22 @@ def forward(self, hidden_states, per_token_scale=None):
...
+                # sparse ULPs, so mirror the FP32 island under the accuracy gate.
+                with paddle.amp.auto_cast(enable=False):
+                    x_glu, x_linear = paddle.chunk(
+                        intermediate_parallel.cast("float32"), 2, axis=-1
+                    )
+                    intermediate_parallel = F.silu(x_glu) * x_linear
+                    if per_token_scale is not None:
+                        scale = per_token_scale.unsqueeze(-1).cast("float32")
+                        intermediate_parallel = intermediate_parallel * scale
+                intermediate_parallel = intermediate_parallel.cast(original_dtype)
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/mlp.py]: P1 **优先级：P1**

`Codestyle Check` 里的 `ruff format` 仍在失败，本地 `ruff format --diff src/paddlefleet/transformer/mlp.py` 定位到这一行需要拆行。当前格式会让必需的样式检查继续失败；请按 formatter 输出调整。

```suggestion
                intermediate_parallel = intermediate_parallel.cast(
                    original_dtype
                )
```

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 5]
```diff
@@ -53,33 +53,35 @@
 from paddlefleet.utils import divide
 
 
+def _use_accuracy_compatible_config(config: "TransformerConfig") -> bool:
+    return bool(getattr(config, "use_accuracy_compatible", False))
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py]: P1 **优先级：P1**

`Codestyle Check` 里 `ruff check` 失败，本地 `ruff check --fix --diff src/paddlefleet/transformer/dot_product_attention.py` 定位到这里的 UP037。该文件已经启用 `from __future__ import annotations`，带引号的类型注解会让样式检查持续失败；请移除注解里的字符串引号。

```suggestion
def _use_accuracy_compatible_config(config: TransformerConfig) -> bool:
    return bool(getattr(config, "use_accuracy_compatible", False))
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。合并冲突已经解除；上一轮 3 个 P1 代码问题在当前代码中仍未修复，继续以原行内线程为准。另发现当前 codestyle 失败中有两个可直接定位的格式/ruff 问题，已补充为新的行内评论。

P3 **优先级：P3**
- 非行级：PR 标题/描述不在 diff 行中，无法行内附着。当前 `Check PR Description` 仍因模板字段缺失失败；建议把标题改成类似 `[Bug fixes] Align Kimi BF16 column linear under accuracy gate`，并在描述里补齐 `PR Category: Distributed Strategy`、`PR Types: Bug fixes`（或更准确的类型）以及现有 Summary/Validation 内容。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by zrr1999** [comment]
Update pushed in `20a36d5`:

- Reset the branch onto latest `develop` and removed the stale broad accuracy/debug diff. The old comments about env mutation, attention, frozen-weight, router/norm/softmax changes, and formatting are now on outdated code that is no longer in this PR.
- Current PR diff is intentionally minimal: `src/paddlefleet/tensor_parallel/layers.py` plus a single-card regression test covering the new gate and default path.
- Updated the PR body to satisfy `ci/checkPRTemplate.py` (`### PR Category`, `### PR Types`, `### Description`).

**Review by risemeup1111 — APPROVED** [review]
已复查最新提交。当前 PR 已重置为最小范围，只保留 `ColumnParallelLinear` experimental 路径下的 BF16 TP=1 accuracy gate 和对应单卡回归测试；上一轮 Nyanpasu 提出的 P1/P3 问题都已随旧的大范围 diff 移除或由当前模板检查通过而关闭。当前代码未发现需要阻塞合入的问题。

本地已验证 `py_compile` 和 `ruff check` 通过；新增测试在本地因环境缺少 `paddle` 无法收集运行，仍以仓库 CI 结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by zrr1999** [comment]
## Kimi-K2 numerical-alignment scope update

This PR should remain narrow: it contributes one measured BF16 column-linear mechanism and keeps the default path unchanged. The results below come from the complete gated Kimi stack, not from this PR alone.

Validated on the real Kimi-K2 2-layer BF16 SFT checkpoint with native ms-swift/Megatron-Core and PaddleFormers/PaddleFleet runners:

- EP1 forward loss: 100/100 steps bitwise exact.
- EP8 all-to-all forward loss: 100/100 steps bitwise exact.
- EP1 step-1 canonical parameter gradients: 794/794 `BITWISE_PASS`, revalidated after updating PFCCLab/Megatron-LM to `a10a87c5b`.
- EP8 bounded canonical gradients: 64/64 `BITWISE_PASS`.
- Bounded optimizer state for two real decay/no-decay parameters: 8/8 exact across FP32 master parameter, FP32 moment1/moment2, and BF16 copyback.

This is **not yet a complete-training alignment claim**. The following native contracts remain to be closed end to end: global gradient norm/clipping, scheduler and warmup authority, all parameter groups/tied weights, distributed optimizer reduction/update order, and multi-step native training updates.

A related authority distinction is important for review. PFCCLab/Megatron-LM PR #2 targets the MiniMax/PyTorch-AdamW stack and cannot be applied wholesale to this Kimi contract:

- Kimi's native optimizer authority is TransformerEngine 2.16.1 `FusedAdam`; forcing `torch.optim.AdamW(fused=True)` changed FP32 master/moment hashes, while the TE-compatible Paddle path is currently 8/8 exact.
- The MiniMax diagnosis in PFCCLab/ms-swift #2 assumes BF16 Megatron main grads versus Paddle `amp_master_grad`; the current Kimi EP8 config already uses FP32 main grads and FP32 grad reduction.
- Replacing native Torch MoE `index_select` backward with FP32 scatter changes the reference semantics. Kimi instead makes Paddle reproduce native deterministic Torch index-add behavior, which is part of the 794/794 result.
- Alternate per-token loss normalization still needs distributed, variable-length, accumulation-aware training evidence before use in Kimi.

Therefore this PR's explicit TP1/BF16/accuracy-gated scope is intentional. It should not absorb model-specific optimizer, MoE-backward, or training-normalization changes.
